### ISSUE TITLE
cant get 97 % accuracy

### ISSUE THREAD

**Issue by Xunius** [makeyourownneuralnetwork/makeyourownneuralnetwork#18]
Hi Tariq,
I watched you "gentle introduction to neural network" video and followed here. I tried the notebook, but I can't get 97% accuracy from the test set. The best I could get is 100% on training set (that needs about 10 epochs of training instead of 5), and 70% on the test set.

Another question is: I've tried to build on your code to make it work for more than 1 hidden layers, using the new version it works just as good as your original code if only 1 hidden layer of 200 nodes is used, works about OK if 2 layers of 200 nodes, but if I added up to 3 layers, the model kind of broke: it always predicts the same digit (e.g. 0) no matter what input is fed in, whether it is a training sample or a test one. I've lowered down the learning rate by several orders of magnitudes and it still behaves like that. Could you give me some hinds why this is happening?

So this is technically not an issue of the code anymore, I couldn't find your email on the internet, and twitter is just too limited to put so many texts, so I tried it here. I've been quite interested in machine learning and neural network, but there isn't nobody around me who could provide me with some guidance, so your help will be much appreciated. 

I'm pasting my code below:
```

import numpy
import matplotlib.pyplot as plt
from scipy.special import expit
import pandas as pd



class neuralNetwork:
    def __init__(self, inputnodes, hiddennodes, outputnodes, learningrate):
        if numpy.isscalar(hiddennodes):
            hiddennodes=[hiddennodes,]
        self.inodes = inputnodes
        self.hnodes = hiddennodes
        self.onodes = outputnodes
        
        self.n_nodes=[inputnodes,]+hiddennodes+[outputnodes,]
        self.n_layers=len(self.n_nodes)

        #-------------Construct weight matrix-------------
        self.thetas=[]

        for ii in range(self.n_layers-1):
            thetaii=numpy.random.normal(0,pow(self.inodes,-0.5),[self.n_nodes[ii+1],self.n_nodes[ii]])
            self.thetas.append(thetaii)
        self.lr = learningrate
        self.activation_function = lambda x: expit(x)
        


    #---------------------Forward---------------------
    def feedForward(self,x,y):
        activations=[x,]
        a1=x
        for jj in range(self.n_layers-1):
            if jj!=self.n_layers-2:
                zjj=numpy.dot(self.thetas[jj],a1)
                ajj=expit(zjj)
                activations.append(ajj)
                a1=ajj
            else:
                zjj=numpy.dot(self.thetas[jj],a1)
                ajj=expit(zjj)
                activations.append(ajj)

        return activations


    #-------------------Compute cost-------------------
    def sampleCost(self,yhat,y):
        jii=-numpy.dot(y,numpy.log(yhat))-numpy.dot((1-y),\
                numpy.log((1-yhat)))
        return jii


    #---------------------Backward---------------------
    def feedBackward(self,activations,y):
        grads=[]
        deltaL=activations[-1]-y

        for jj in range(self.n_layers-2,-1,-1):
            deltajj=numpy.dot(self.thetas[jj].T,deltaL)*activations[jj]*\
                    (1-activations[jj])
            gradjj=numpy.outer(deltaL,activations[jj])
            grads.append(gradjj)
            deltaL=deltajj

        grads=grads[::-1]
        return grads

    #-----------------Gradient descent-----------------
    def gradientDescent(self,devs):
        for l in range(self.n_layers-1):
            thetal=self.thetas[l]
            thetal=thetal-self.lr*devs[l]
            self.thetas[l]=thetal
        return


    def train2(self,inputs_list,targets_list):
        inputs=inputs_list
        targets=targets_list

        activations=self.feedForward(inputs,targets)
        j=self.sampleCost(activations[-1],targets)
        grads=self.feedBackward(activations,targets)
        self.gradientDescent(grads)

        return j

    def predict(self,x):
        a1=x
        for jj in range(self.n_layers-1):
            if jj!=self.n_layers-2:
                zjj=numpy.dot(self.thetas[jj],a1)
                ajj=expit(zjj)
                a1=ajj
            else:
                zjj=numpy.dot(self.thetas[jj],a1)
                ajj=expit(zjj)
        return ajj



def readData(path):
    data_file=pd.read_csv(path,header=None)
    x_data=[]
    y_data=[]
    for ii in xrange(len(data_file)):
        lineii=data_file.iloc[ii,:]
        imgii=numpy.array(lineii.iloc[1:])/255.*0.99+0.01
        labelii=numpy.zeros(10)+0.01
        labelii[int(lineii.iloc[0])]=0.99

        y_data.append(labelii)
        x_data.append(imgii)

    x_data=numpy.array(x_data)
    y_data=numpy.array(y_data)

    return x_data,y_data




if __name__=='__main__':

    #-----------------Open MNIST data-----------------
    path='/home/guangzhi/codes/nn/mnist_data/mnist_train_100.csv'
    x_data,y_data=readData(path)

    #------------------Open test data------------------
    path='/home/guangzhi/codes/nn/mnist_data/mnist_test_10.csv'
    x_data_test,y_data_test=readData(path)

    #------------------Create network------------------
    ninnodes=784
    nhidden=[200,200,100]
    noutnodes=10
    lr=0.1
    nn=neuralNetwork(ninnodes,nhidden,noutnodes,lr)

    #----------------------Train----------------------
    costs=[]
    epoch=60
    for ii in range(epoch):
        idxs=numpy.random.permutation(len(x_data))
        for jj in idxs:
            jjj=nn.train2(x_data[jj],y_data[jj])
            costs.append(jjj)
            print jjj

    #----------------Plot cost function----------------
    plt.plot(costs);plt.show(block=False)

    #---------------Validate on train set---------------
    n_test=100
    n_correct=0
    for ii in range(n_test):
        yhat=numpy.argmax(nn.predict(x_data[ii]))
        yii=numpy.argmax(y_data[ii])
        if yhat==yii:
            n_correct+=1
        print 'yhat = %d, yii = %d' %(yhat, yii)

    print n_correct/float(n_test)*100.

    #---------------Validate on test set---------------
    n_test=10
    n_correct=0
    for ii in range(n_test):
        yhat=numpy.argmax(nn.predict(x_data_test[ii]))
        yii=numpy.argmax(y_data_test[ii])
        if yhat==yii:
            n_correct+=1
        print 'yhat = %d, yii = %d' %(yhat, yii)

    print n_correct/float(n_test)*100.
    

```

**Comment by Xunius** [comment]
Reply to myself: I've figured out the issue, I need to use the entire dataset of mnist rather than a small section of it for the training, and also the complete testing set.
3-layered network does converge, but takes a longer time, probably because gradient descent has fallen into a narrow tunnel.
I've also added a regularization term to counter overfitting. After some try-n-error I managed to get 98.2% on the test set.
Hope to see more nice demonstrations like this, it's reallly helpful!

**Comment by makeyourownneuralnetwork** [comment]
thanks for getting in touch and sorry for the delay in getting back to you .. I have been a little busy with life (moving home, jobs etc)

what you have found is what many people also find - that deeper networks take longer to train.

thanks also for clarifying and sharing your experience - e.g. using the full MNIST dataset .. others reading this will find it useful. 

if you do an interesting project - I am happy to share it from the book's blog, which already has some interesting guest posts or mentions of interesting projects inspired by the book
[https://makeyourownneuralnetwork.blogspot.co.uk](https://makeyourownneuralnetwork.blogspot.co.uk)
